### DISCUSSION TITLE
py5 format error at properties: py5.height and py5.width

### DISCUSSION THREAD

**Discussion by claudefalbriard** [py5coding/py5generator#49] [Q&A] [answered]
I'm new to the py5 project effort, so first a welcome to all members.
My environment is: Linux Ubuntu 18, Anaconda, Python 3.8.5 and Jupyter 
When calling py5.height, or py5.width in a sketch, I get a method in place of the expected numeric value.  The data returned is: 
`<bound method Sketch._get_height of <py5.sketch.Sketch object at 0x7f41722c98e0>>`
When calling it as a method, like py5.height() its going well, or by using the attributes "height", or "width" with no prefix, also  works OK, as expected.  Wonder if this is a planned behavior for the two properties.

**Comment by hx2A** [comment, marked as answer]
Welcome, @claudefalbriard !

Can I see your code? Also, how are you running this? Are you running it using the py5 kernel or using run_sketch?

Are you coding in "imported mode" or "module mode"? It is not a good idea to mix the the two together in the same sketch. I am not sure what the result might be but I suspect it might cause those symptoms.

**Reply by claudefalbriard** [reply]
Thanks for your analysis and assistance. I am using the jupyter "p5" mode. Looks like my code runs in default "Module Mode".  Below my code to simulate the issue: 
```
# discussion: https://github.com/hx2A/py5generator/discussions/49

import py5

print("dir py----------------")
print(dir(py5))
print("dir py5.sketch.Sketch-")
print(dir(py5.sketch.Sketch))
print("----------------------")
      
particles = []

vs = 400
hs = 400

def setup():
    py5.background(0)
    py5.smooth()
    py5.stroke_weight(4)

# cfa quick fix         
def settings():
    py5.size(vs, hs) 
    
def draw():
    py5.no_stroke()    

print(py5.height) # throws an error
print(py5.width)  # throws an error

print(py5.height()) # read with a method
print(py5.width())  # read with a method

print(height) # reads py5.sketch.Sketch attribute
print(width)  # reads py5.sketch.Sketch attribute

py5.run_sketch()

Log: 
----------------------
<bound method Sketch._get_height of <py5.sketch.Sketch object at 0x7fee557638e0>>
<bound method Sketch._get_width of <py5.sketch.Sketch object at 0x7fee557638e0>>
100
100
100
100
```

**Reply by hx2A** [reply]
OK, I see what the problem is. First, some background:

http://py5.ixora.io/content/py5_modes.html

The height and width values are attributes of the `py5.Sketch` class and are implemented as properties. When you program in "class mode" (see the above link), you are using these properties to get the height and width values.

When you program in module mode, the code you write to get the height and width is `py5.height` and `py5.width`. The py5 module actually has no height or width properties, but they work because of a module level `__getattr__` function essentially redirects the attribute request to the current sketch. If it wasn't for [PEP 562](https://www.python.org/dev/peps/pep-0562/) module mode wouldn't work at all.

When you program in imported mode, the code you write to get the height and width is `height` and `width`. In this case there can't be a module level `__getattr__` function to link to the Sketch instance's property. Instead, imported mode links those names to the `_get_height` and `_get_width` methods in the Sketch class (these are the getters for the property). But linking to methods is not the same as calling them. To make it all work, there needs to be an AST transformation that essentially searches for `height` and `width` and adds a `()` at the end. This is done after your code is compiled but before execution. Conveniently, the IPython Kernel allows you to add custom AST transformations if you want to create your own custom kernel, which is what the py5 kernel is, and is what you are using here.

Now, for what you are doing:

If you are using the py5 kernel, you are programming in "imported mode". You shouldn't import py5 because that is done for you, and it will cause problems if you try to mix imported mode code with module mode code. Imported mode is more convenient because you don't have to keep typing `py5.` all over the place.

If you use the regular Python kernel, you are programming in "module mode". You do have to import py5 and you will need to prefix the py5 functions and variables like `height` and `width` with `py5.`

To move forward, I recommend you either remove the `py5.` prefixes or switch to the regular Python kernel. In Jupyter notebook or Jupyter lab you can switch kernels by clicking on the kernel name in the upper right corner.

**Reply by hx2A** [reply]
BTW, if you like you can put the call to `size()` in the beginning of `setup()`, before the other function calls. You don't need `settings()` anymore. This is similar to programming in Java in the Processing IDE.

**Reply by hx2A** [reply]
Also, I read this line in your code:

```
particles = []
```

and am curious what you are working on. This might not be relevant, but there is a new `points()` method that lets you draw multiple points instead of calling `point()` repeatedly in a loop.

**Reply by claudefalbriard** [reply]
Thanks so much for the detailed clarification.
 
I really was unaware about the details and differentiated use of the Jupyter P5 kernel and import modes. The use of the P5 kernel simplifies and shortens the code. There might exist more complex code samples where we need to add extensions by Python 3 classes, like the "Particles" and justifies the use of module mode and standard Python 3 kernels, addressing classes and functions. The conflict with the size parameter and use of settings() function had the same origin. 

I will keep this discussion in my Jupyter sample code for future reference.

**Reply by hx2A** [reply]
Sounds great. When you are using the py5 kernel you can import any other Python library you want. It really is just the IPython kernel with a few small customizations.

I'll see what I can do about adding a warning when users execute `import py5` from within the py5 kernel. That might be possible, and will help others who run into this same issue.

**Reply by hx2A** [reply]
> I'll see what I can do about adding a warning when users execute `import py5` from within the py5 kernel. That might be possible, and will help others who run into this same issue.

Done. Now the py5 kernel (and py5bot, and the run_sketch terminal command) will reject code that tries to execute `import py5`. Users will see a message like this:

```
"import py5" found on line 1. Do not import the py5 library, as this has already been done for you. Your code should be written without any "py5." prefixes.
```
